src/rsync.rs

```diff
@@ -1,76 +1,259 @@
 }
 
-pub fn build_repo_from_rrdp_snapshot(
-    opt: &Opt,
-    notify: &mut NotificationFile,
-    client: &HttpClient,
-    raw_snapshot: &[u8],
-    last_serial: RrdpSerialNumber,
+pub fn update_from_rrdp_state(
+    rrdp_state: &RrdpState,
+    changed: bool,
+    config: &Config,
 ) -> Result<()> {
-    info!("Updating Rsync repository");
+    let mut rsync_state = RsyncDirState::recover(config)?;
 
-    let out_path = if cfg!(unix) {
-        let extension = format!("{}_{}", last_serial, Utc::now().timestamp());
-        file_ops::set_path_ext(&opt.rsync_dir, &extension)
-    } else {
-        file_ops::set_path_ext(&opt.rsync_dir, config::TMP_FILE_EXT)
+    let new_revision = RsyncRevision {
+        session_id: rrdp_state.session_id(),
+        serial: rrdp_state.serial(),
     };
 
+    if changed {
+        write_rsync_content(&new_revision.path(config), rrdp_state.elements())?;
+
+        if config.rsync_dir_use_symlinks() {
+            symlink_current_to_new_revision_dir(&new_revision, config)?;
+        } else {
+            rename_new_revision_dir_to_current(&new_revision, &rsync_state, config)?;
+        }
+
+        rsync_state.update_current(new_revision);
+    }
+
+    rsync_state.clean_old(config)?;
+    rsync_state.persist(config)?;
+
+    Ok(())
+}
+
+/// Create a new symlink then rename it. We need to do this because the std library
+/// refuses to overwrite an existing symlink. And if we were to remove it first, then
+/// we would introduce a race condition for clients accessing.
+fn symlink_current_to_new_revision_dir(
+    new_revision: &RsyncRevision,
+    config: &Config,
+) -> Result<()> {
     info!(
-        "Writing Rsync repository to: {}",
-        out_path.to_string_lossy()
+        "Updating symlink 'current' to '{}' under rsync dir '{:?}'",
+        new_revision.dir_name(),
+        config.rsync_dir
     );
-    write_rsync_content(&out_path, notify, client, raw_snapshot)?;
-
-    if cfg!(unix) {
-        info!("Use symlink to link rsync module dir to the new content");
-        // create a new symlink then rename it
-        let tmp_name = file_ops::set_path_ext(&opt.rsync_dir, config::TMP_FILE_EXT);
-        std::os::unix::fs::symlink(&out_path, &tmp_name)?;
-        std::fs::rename(&tmp_name, &opt.rsync_dir)?;
-    } else {
-        info!("Renaming rsync folders for close to atomic update of the rsync module dir");
-        file_ops::install_new_dir(&opt.rsync_dir, last_serial.to_string())?;
+    let current_path = config.rsync_dir_current();
+
+    let tmp_name = file_ops::path_with_extension(&current_path, config::TMP_FILE_EXT);
+    if tmp_name.exists() {
+        std::fs::remove_file(&tmp_name).with_context(|| {
+            format!(
+                "Could not remove lingering temporary symlink for current rsync dir at '{:?}'",
+                tmp_name
+            )
+        })?;
     }
 
+    std::os::unix::fs::symlink(new_revision.dir_name(), &tmp_name).with_context(|| {
```

Comment:
I think it would be good to say here that we link then rename because it's not permitted to create a link directly over an existing link. I was wondering why not just do this in one step without the rename but when I tried I see that you get an already exists error.

Reply:
See line 58:

/// Create a new symlink then rename it. We need to do this because the std library
/// refuses to overwrite an existing symlink. And if we were to remove it first, then
/// we would introduce a race condition for clients accessing.